component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,236 @@
+    public T computeIfAbsent(String key,
+            final Predicate<Element<T>> toRemove,
+            final Supplier<T> value) {
+
+        final ElementImpl<T> element = this.addToMap(key,
+                () -> new ElementImpl<>(value, toRemove, -1, null));
+
+        return element.value;
+    }
+
+    @Override
+    public T computeIfAbsent(String key, long timeoutMs, Supplier<T> value) {
+
+        final long endOfValidity = timeoutMs > 0 ? System.currentTimeMillis() + timeoutMs : -1;
+        final ScheduledFuture<?> task = endOfValidity > 0 ? evictionTask(key, endOfValidity) : null;
+
+        final ElementImpl<T> element = addToMap(key,
+                () -> new ElementImpl<T>(value, null, endOfValidity, task));
+
+        return element.value;
+    }
+
+    private ElementImpl<T> addToMap(String key, Supplier<ElementImpl<T>> builder) {
         final String internalKey = internalKey(key);
-        final Element element = cache
-                .compute(internalKey,
-                        (k, e) -> e == null || e.isExpired()
-                                ? new Element(value.get(), System.currentTimeMillis() + timeoutMs)
-                                : e);
-        if (element == null) {
-            return null;
-        }
-        return (T) element.value;
+        return cache.compute(
+                internalKey,
+                (String k, ElementImpl<T> old) ->    //
+                        old == null || old.mustBeRemoved() ?
+                                builder.get() :
+                                old
+        );
+    }
+
+    @Override
+    public T computeIfAbsent(String key, Supplier<T> value) {
+        final CacheConfiguration config = this.configuration.get();
+        long timeOut = config.isActive() ? config.getDefaultEvictionInterval() : -1;
+        return computeIfAbsent(key, timeOut, value);
     }
 
     private String internalKey(final String key) {
         return plugin + '@' + key;
     }
 
-    @AllArgsConstructor
-    public static class Element {
+    public void clean() {
+        final List<String> removableElements = this.cache
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue().mustBeRemoved())
+                .map(Entry::getKey)
+                .collect(toList());// materialize before actually removing it
+        removableElements.forEach(this.cache::remove);
+    }
+
+    public boolean isEmpty() {
+        return this.cache.isEmpty();
+    }
+
+    private ScheduledExecutorService getThreadService() {
+        if (this.threadService == null) {
+            synchronized (this) {
+                if (this.threadService == null) {
+                    this.threadService = Executors.newSingleThreadScheduledExecutor((Runnable r) -> {
+                        final Thread thread = new Thread(r, this.getClass().getName() + "-eviction-" + hashCode());
```

Comment:
this getClass() will not be the one you want (likely the lambda one), maybe reference the LocalCacheService.class.getName() instead, it should also get the plugin to differenciate instances per family

side note: best would be to have a single global default evictor when config is the same to avoid to create an uncontrolled number of thread (x#families which can be a lot in the server) IMO

Reply:
Number of thread is defined by threadService class :
this.threadService = Executors.newScheduledThreadPool(4, (Runnable r) -> ..) 
will never go over 4 thread, only the Queue that will have numbers of Runnable.

Reply:
do you think? this number is per family using local cache (so almost all) so it will quickly grow if not factorised with a default global pool (lazily created too) when no specific configuration is set.

Reply:
So, PreDestroy should also be for this global pool (shutdownNow & awaitTermination).
How to have one global pool for several LocalCache (@ Inject => but not lazy, @ Service, Add to constructor or static instance ?)